### PR TITLE
WIP: support clusters

### LINKED ISSUE #38: Clusters are not supported
since the latest version i get the error:
```
2017/09/13 13:42:55 [INFO] (telemetry) found error: path '/Interxion MAD2/host/cluster1' resolves to multiple hosts
1505310175,,ui,error,Build 'vsphere' errored: path '/Interxion MAD2/host/cluster1' resolves to multiple hosts
```
seems to be an issues with the `HostSystem` option as this does such checks
https://github.com/jetbrains-infra/packer-builder-vsphere/blob/a12121616df50a6e1d0dfd570620e132048b7771/driver/host.go#L23

it worked on `v.1.3.1.dev1` & `v.1.3`

maby a possibility to release `1.3.1`?? as it seems you have removed ` v.1.3.1.dev1`

### PR DESCRIPTION
**Pull request by amarruedo**
This PR fixes #38 

First, I have to say I'm not go developer, nor have much expertise in vSphere resource managing, but I needed this solved. 

All I have done is have a look on how do guys from [terraform-vsphere-provider](https://github.com/terraform-providers/terraform-provider-vsphere) do to discover `resource_pools` and `esxi hosts`, so all the credit goes to them.

So far the tests pass, and I also have been able to create a virtual machine on my vSphere cluster.

Probably there is a better way to achieve this, or there is some corner case I totally didn´t take in account, so I hope this PR inspires the author of this packer builder, or some other developer that is more skilled, to create a proper fix.

### CODE REVIEW

**Comment by seanmalloy** [comment]
@amarruedo thanks for submitting this PR! I need this feature too. I'll try to test out this patch in my vCenter 6.0 environment in the next week and provide some feedback.

**Comment by jwhitesoe** [comment]
I'd like to look at this.  I'm on windows.  Do I need to build this on windows or can I use the linux subsystem?

I'm pretty go-clueless.

**Comment by seanmalloy** [comment]
@jwhitesoe I am running in Linux and I was able to build the Linux, Mac  OSX, and Windows binaries. I'm assuming you would be able to do the same on Windows.

**Comment by seanmalloy** [comment]
I tested this patch using both the `vsphere-iso` and `vsphere-clone` builders. Both of them work in my vCenter/ESXi 6.0 environment. All of the ESXi hosts in my environment are in clusters(no stand alone hosts). 

No magic Packer JSON template changes required. Simply specify a `host` that is part of an ESXi cluster and it just works.

Example configs used to test.

```
{
  "builders": [
    {
      "type": "vsphere-iso",
      "vcenter_server": "vcenter.example.com",
      "username": "myvcenteruser",
      "password": "myvcenterpass",
      "ssh_username": "root",
      "ssh_password": "myrootpass",
      "insecure_connection": "true",
      "vm_name": "packer-example",
      "host": "myesxhost.example.com",
      "datastore": "MY-DATASTORE-FULL",
      "network": "10.7.142.0_24",
      "guest_os_type": "rhel7_64Guest",
      "CPUs": 1,
      "RAM": 4096,
      "disk_controller_type": "pvscsi",
      "disk_size": 300,
      "disk_thin_provisioned": true,
      "network_card": "vmxnet3",
      "iso_paths": [
        "[TEMPLATES] install_media/rhel-server-7-install-20180311195005.iso"
      ],
      "boot_order": "disk,cdrom",
      "convert_to_template": true
    }
  ],
  "provisioners": [
    {
      "type": "shell",
      "inline": ["ls /"]
    }
  ]
}
```

```
{
  "builders": [
    {
      "type": "vsphere-clone",
      "vcenter_server": "vsphere.example.com",
      "username": "myvcenterusername",
      "password": "myvcenterpass",
      "ssh_username": "root",
      "ssh_password": "myrootpass",
      "insecure_connection": "true",
      "template": "packer-example",
      "vm_name": "packer-clone-example",
      "host": "myesxhost.example.com",
      "datastore": "MY-DATASTORE-FULL",
      "CPUs": 1,
      "RAM": 4096,
      "disk_size": 300,
      "convert_to_template": true
    }
  ],
  "provisioners": [
    {
      "type": "shell",
      "inline": ["ls /"]
    }
  ]
}
```

**Comment by seanmalloy** [comment]
It would be great is someone could test this patch against an ESXi host that is NOT part of a cluster to verify that this patch does not break existing functionality. I'm not able to do this because all of the hosts in my environment at part of clusters.

Also, please let me know if any additional work is need to get this patch merged(more testing, docs, code changes, etc.). I'm willing to help with whatever is needed.

Thanks!

**Comment by amarruedo** [comment]
@seanmalloy thanks for testing this. All my hosts are part of a cluster too so I can't verify whether or not the patch breaks existing functionallity.

**Comment by amarruedo** [comment]
@mkuzmin I just made a minor change, got rid of `host` variable in `FindResourcePool`function since it is no longer needed.

**Comment by mkuzmin** [comment]
@jwhitesoe you can use Docker: `docker-compose run build`

**Comment by jwhitesoe** [comment]
@mkuzmin No such luck on windows.  I had to update the docker-compose.yml to support version 2.1

version: '2.1'
services:
    build:
        image: jetbrainsinfra/golang:1.9.4
        volumes:
            - .:/go/src/github.com/jetbrains-infra/packer-builder-vsphere
        working_dir: /go/src/github.com/jetbrains-infra/packer-builder-vsphere
        command: ./build.sh

But then I run into this error (which seems like an unfixed bug with glide and windows) :

[ERROR] Unable to strip vendor directories: remove /go/src/github.com/jetbrains-infra/packer-builder-vsphere/vendor/github.com/hashicorp/packer/vendor/github.com/jmespath/go-jmespath/fuzz/corpus: directory not empty

**Comment by jwhitesoe** [comment]
Trying to build on the WSL on win10 gets me here :

(This might be a $GOROOT issue?)

[INFO]  Removing: /root/work/src/github.com/packer-builder-vsphere/packer-builder-vsphere/vendor/github.com/hashicorp/go-multierror/vendor
[INFO]  Removing: /root/work/src/github.com/packer-builder-vsphere/packer-builder-vsphere/vendor/github.com/hashicorp/packer/vendor
[INFO]  Removing: /root/work/src/github.com/packer-builder-vsphere/packer-builder-vsphere/vendor/github.com/vmware/govmomi/vendor
+ export CGO_ENABLED=0
+ export GOARCH=amd64
+ mkdir -p bin
+ rm -f bin/*
+ GOOS=darwin go build -o bin/packer-builder-vsphere-iso.macos ./cmd/iso
cmd/iso/main.go:3:8: cannot find package "github.com/hashicorp/packer/packer/plugin" in any of:
        /usr/lib/go-1.6/src/github.com/hashicorp/packer/packer/plugin (from $GOROOT)
        /root/work/src/github.com/packer-builder-vsphere/packer-builder-vsphere/src/github.com/hashicorp/packer/packer/plugin (from $GOPATH)
cmd/iso/main.go:4:8: cannot find package "github.com/jetbrains-infra/packer-builder-vsphere/iso" in any of:
        /usr/lib/go-1.6/src/github.com/jetbrains-infra/packer-builder-vsphere/iso (from $GOROOT)
        /root/work/src/github.com/packer-builder-vsphere/packer-builder-vsphere/src/github.com/jetbrains-infra/packer-builder-vsphere/iso (from $GOPATH)
root@Jeremy-PC:~/work/src/github.com/packer-builder-vsphere/packer-builder-vsphere# go build -o bin/packer-builder-vsphere-iso.exe   ./cmd/iso
cmd/iso/main.go:3:8: cannot find package "github.com/hashicorp/packer/packer/plugin" in any of:
        /usr/lib/go-1.6/src/github.com/hashicorp/packer/packer/plugin (from $GOROOT)
        /root/work/src/github.com/packer-builder-vsphere/packer-builder-vsphere/src/github.com/hashicorp/packer/packer/plugin (from $GOPATH)
cmd/iso/main.go:4:8: cannot find package "github.com/jetbrains-infra/packer-builder-vsphere/iso" in any of:
        /usr/lib/go-1.6/src/github.com/jetbrains-infra/packer-builder-vsphere/iso (from $GOROOT)
        /root/work/src/github.com/packer-builder-vsphere/packer-builder-vsphere/src/github.com/jetbrains-infra/packer-builder-vsphere/iso (from $GOPATH)

**Comment by jwhitesoe** [comment]
@amarruedo I got this compiled by creating a ubuntu 16 vm on hyper-v, then installing docker and building there.

I am successfully provisioning on our vsphere with clusters using ""type": "vsphere-iso",

I have a second order problem where the disk configuration is incompatible with my normal windows autounattend, but this seems like a second order problem.

So far, this pull request looks very good.

**Comment by jwhitesoe** [comment]
The packer iso image is started and the windows autounattend.xml file is running successfully.

What is not happening is packer is still unable to connect to the started machine after this runs.

What I see:

==> vsphere-iso: Creating VM...
==> vsphere-iso: Adding CDRoms...
==> vsphere-iso: Adding configuration parameters...
==> vsphere-iso: Power on VM...
==> vsphere-iso: Waiting 10s for boot...
==> vsphere-iso: Typing boot command...
==> vsphere-iso: Waiting for IP...

My packer.json file is here: 

``{
  "builders": [
    {
      "type": "vsphere-iso",
      "vm_name":  "vm-1",

      "vcenter_server": "sa-vcentre.xxxxxx.com",
      "insecure_connection": "true",
      "username": "xxxxx@xxxxxx",
      "password": "xxxxxx",

      "datacenter" : "SOE Auckland",
      "host": "sa-esxi-1.xxxxxx.com",
    
      "datastore" : "SOE AKL LAB SAS Volume 01",

      "CPUs": "2",
      "RAM": "4096",

      "guest_os_type": "windows9_64Guest",
      "disk_controller_type" : "lsilogic-sas",
      "disk_size": "32",
 
      "network_card": "e1000e",
      "network" : "VLAN65 - 10.240.65.0 Server",

      "communicator": "winrm",
      "winrm_username": "vagrant",
      "winrm_password": "vagrant",

      "shutdown_command": "shutdown /s /t 10 /f /d p:4:1 /c \"Packer Shutdown\"",
      "shutdown_timeout": "15m",

      "iso_paths": [
        "[SOE AKL LAB SAS Volume 01] Win10_for_vsphere.iso"
      ],


      "convert_to_template": "true",
      "folder": "Templates"

    }
  ],
  "provisioners": [
    {
        "type": "powershell",
        "inline":[
            "Write-Host foo"
        ]
    }
]


  
}

**Comment by seanmalloy** [comment]
@jwhitesoe I had a similar issue this first time I tried the vsphere-iso builder, but I was using Linux. The automated RHEL install process had to be configured to reboot the VM after the installer was done. I have no idea how to do that with Windows, but I think that is your problem. Also make sure that VMWare tools gets installed.

It might be helpful to look over the below file to see what you might be missing.

https://github.com/jetbrains-infra/packer-builder-vsphere/blob/master/examples/windows/setup/Autounattend.xml

**Comment by xjsp** [comment]
Have been watching this for a bit and can confirm this works well when explicitly declaring the Host inside a cluster
```
{
  "builders": [
    {
      "type": "vsphere-iso",
      "vcenter_server": "x.x.net",
      "username": "x",
      "password": "x",
      "ssh_username": "root",
      "ssh_password": "x",
      "insecure_connection": "true",
      "vm_name": "rhel7.4-current-cloud",
      "folder": "Backup_Groups/DC1/DC1_Templates",
      "host": "/DC/host/lux-san-dc1/evmhvi000209.x.net",
      "datastore": "xpaa_vi_027",
      "resource_pool": "/DC/host/lux-san-dc1/Resources",
      "datacenter": "DC",
      "network": "LAB-CHDC",
      "guest_os_type": "rhel7_64Guest",
      "CPUs": 1,
      "RAM": 1024,
      "disk_controller_type": "pvscsi",
      "disk_size": 16,
      "disk_thin_provisioned": true,
      "network_card": "vmxnet3",
      "iso_paths": [
        "[xpaa_vi_033] lux_templates/rhel-server-7.4-x86_64-dvd.iso"
      ],
      "boot_order": "cdrom,disk",
      "convert_to_template": true
    }
  ],
  "provisioners": [
    {
      "type": "shell",
      "inline": ["ls /"]
    }
  ]
}
```
(The host value could have been specified with only 'evmhvi000209.x.net' instead of the full path, done for demonstration purposes )
Some specifics on the vSphere env: Consists of Hosts in clusters, datastores in datastoreclusters and no custom Resource Pools.
Not sure what other work needs to be done in order to enable a "cluster" only parameter where vSphere takes care of the placement via DRS (same for datastoreclusters). But even this PR is good as is 👍 Thanks @amarruedo :)

File: driver/resource_pool.go [hunk 1]
```diff
@@ -19,15 +20,44 @@ func (d *Driver) NewResourcePool(ref *types.ManagedObjectReference) *ResourcePoo
...
+			p, err := d.finder.ResourcePool(d.ctx, name)
+			if err != nil {
+				return nil, err
+			}
+			return &ResourcePool{
+				pool:   p,
+				driver: d,
+			}, nil
+		}
+		p, err := d.finder.DefaultResourcePool(d.ctx)
```
Comment (mkuzmin) [inline driver/resource_pool.go]: I have multiple resource pools , so it fails with `Build 'vsphere-iso' errored: error creating vm: default resource pool resolves to multiple instances, please specify`

File: driver/resource_pool.go [hunk 2]
```diff
@@ -19,15 +20,44 @@ func (d *Driver) NewResourcePool(ref *types.ManagedObjectReference) *ResourcePoo
...
+		if err != nil {
+			return nil, err
+		}
+		return &ResourcePool{
+			pool:   p,
+			driver: d,
+		}, nil
+	case "VirtualCenter":
+		if name != "" {
+			p, err := d.finder.ResourcePool(d.ctx, name)
```
Comment (mkuzmin) [inline driver/resource_pool.go]: this changes a syntax. previously pools name were relative to a host. Now a full path is required: 
```
esxi-1.vsphere65.test/Resources
esxi-1.vsphere65.test/Resources/pool1
cluster1/Resources
cluster2/Resources
cluster2/Resources/pool1
```

**Comment by mkuzmin** [comment]
I see next scenarios:
1) A standalone host, root resource pool
```
"host": "esxi-1.vsphere65.test",
```
2) A standalone host, nested resource pool
```
"host": "esxi-1.vsphere65.test",
"resource_pool": "pool1",
```
3) Cluster without DRS
```
"host": "esxi-2.vsphere65.test",
```
4) Cluster with DRS, root resource pool
```
"cluster": "cluster2",
```
5) Cluster with DRS, nested resource pool
```
"cluster": "cluster2",
"resource_pool": "pool1",
```

Maybe DRS-enabled clusters require a special syntax.

**Comment by seanmalloy** [comment]
@amarruedo do you need help making these code changes based on @mkuzmin's suggestions?

**Comment by mkuzmin** [comment]
@amarruedo thank you again for the help and research here.
As you said above, this is a kind of an issue where an alternate PR would work better.
Please review #61 and let me know what do you think.

**Comment by jwhitesoe** [comment]
@seanmalloy Everything is working now on my end.  Changing how vmware tools was installed allowed the IP to be discovered.  Thanks for the help.

File: driver/resource_pool.go (hunk 2 above)
Comment (amarruedo) [reply driver/resource_pool.go]: I can't try this by myself (my cluster has no DSR and I can't create `resource_pools`) but checking the [code](https://github.com/vmware/govmomi/blob/master/find/finder.go#L845) I think that you can either provide only the name of the `resource_pool`or the complete path.

In the examples you provide you could use only the name of the pool, considering there exist a unique `resource_pool`with that name in your vCenter, otherwise, specifying only the name `pool1`, you would get the error:

```
==> vsphere-iso: Creating VM...
Build 'vsphere-iso' errored: error creating vm: default resource pool resolves to multiple instances, please specify
```

In this case, you would have to use the complete path of the `resource_pool`, i.e. `cluster2/Resources/pool1`

This same thing is happening when trying to find de default `resource_pool`. Checking the [code](https://github.com/vmware/govmomi/blob/master/find/finder.go#L858), we are searching for  `*/Resources`, the root `resurce_pool`. In your example, `DefaultResourcePool`would fail again with the previous error, because we have 3 different root `resource_pools`on the vCenter (`esxi-1.vsphere65.test/Resources`, `cluster1/Resources` and `cluster2/Resources`) . Again, in this case we would have to use the complete path of the `resource_pool`.

I don't see this a problem, as long as your pool name is unique, you can use only the name; if your vCenter has only one root `resource_pool`, you can leave it blank, and in other cases (repeated pool name or more than one root `resource_pool`), you will have to specify the complete path.

File: driver/host.go [hunk 3]
```diff
@@ -1,36 +1,61 @@
...
 	}
 }
 
 func (d *Driver) FindHost(name string) (*Host, error) {
-	h, err := d.finder.HostSystem(d.ctx, fmt.Sprintf("/%v/host/%v", d.datacenter.Name(), name))
-	if err != nil {
-		return nil, err
+	t := d.client.ServiceContent.About.ApiType
+	switch t {
+	case "HostAgent":
```
[the remark this answers is not in the document]
Comment (amarruedo) [reply driver/host.go]: ok then, vCenter only

File: driver/resource_pool.go (hunk 1 above)
Comment (amarruedo) [reply driver/resource_pool.go]: As explained in the previous comment, if you have more than one root `resource_pool` or `resource_pools` with the same name (and declare them on the config only by the name, not the full path), this will happen. This just means that you must be a little more specific and setting the complete path to the desired `resource_pool` would solve the issue.

**Comment by mkuzmin** [comment]
@amarruedo Asier, Thank you for the research and sorry I haven't accepted this PR as is.
That was the case where reviews and fixes would take much more time for both of us.

**Comment by amarruedo** [comment]
No worries :-)
